yjit/src/asm/x86_64/mod.rs

```diff
@@ -617,33 +624,74 @@ impl CodeBlock
         cb->num_refs = 0;
     }
     */
-}
 
-/*
-// Check if an operand needs a REX byte to be encoded
-static bool rex_needed(x86opnd_t opnd)
-{
-    if (opnd.type == OPND_NONE || opnd.type == OPND_IMM)
-    {
-        return false;
+    fn mark_position_writeable(&mut self, write_pos: usize) {
+        let pagesize = unsafe { libc::sysconf(libc::_SC_PAGESIZE) as usize };
+        let aligned_position = (self.write_pos / pagesize) * pagesize;
+
+        if self.current_aligned_write_pos != aligned_position {
+            self.current_aligned_write_pos = aligned_position;
+            let page_addr = self.get_ptr(aligned_position) as *mut libc::c_void;
+
+            unsafe {
+                libc::mprotect(page_addr, pagesize, libc::PROT_READ | libc::PROT_WRITE);
+            }
+        }
     }
 
-    if (opnd.type == OPND_REG)
-    {
-        return (
-            opnd.as.reg.reg_no > 7 ||
-            (opnd.num_bits == 8 && opnd.as.reg.reg_no >= 4 && opnd.as.reg.reg_no <= 7)
-        );
+    /// Write the REX byte
+    fn write_rex(&mut self, w_flag: bool, reg_no: u8, idx_reg_no: u8, rm_reg_no: u8) {
+        // 0 1 0 0 w r x b
+        // w - 64-bit operand size flag
+        // r - MODRM.reg extension
+        // x - SIB.index extension
+        // b - MODRM.rm or SIB.base extension
+        let w: u8 = if w_flag { 1 } else { 0 };
+        let r: u8 = if (reg_no & 8) > 0 { 1 } else { 0 };
+        let x: u8 = if (idx_reg_no & 8) > 0 { 1 } else { 0 };
+        let b: u8 = if (rm_reg_no & 8) > 0 { 1 } else { 0 };
+
+        // Encode and write the REX byte
+        self.write_byte(0x40 + (w << 3) + (r << 2) + (x << 1) + (b));
     }
 
-    if (opnd.type == OPND_MEM)
-    {
-        return (opnd.as.mem.base_reg_no > 7) || (opnd.as.mem.has_idx && opnd.as.mem.idx_reg_no > 7);
+    /// Write an opcode byte with an embedded register operand
+    fn write_opcode(&mut self, opcode: u8, reg: &X86Reg) {
+        let op_byte: u8 = opcode | (reg.reg_no & 7);
+        self.write_byte(op_byte);
     }
```

Comment:
These probably should go on your `Assembler` trait rather than CodeBlock since they're not portable.

We might want to also move `CodeBlock` to a separate file, like maybe `src/asm/mod.rs`. We should separate out the code that's not specific to x86 so that we can reuse it for arm.